nuxeo-coldstorage/src/test/java/org/nuxeo/coldstorage/operations/AbstractTestRestoreFromColdStorageOperation.java

```diff
@@ -0,0 +1,107 @@
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.event.EventService;
+import org.nuxeo.ecm.platform.ec.notification.NotificationConstants;
+import org.nuxeo.ecm.platform.notification.api.NotificationManager;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+
+public abstract class AbstractTestRestoreFromColdStorageOperation extends AbstractTestColdStorageOperation {
+
+    @Inject
+    protected CoreSession session;
+
+    @Inject
+    protected NotificationManager notificationManager;
+
+    @Inject
+    protected EventService eventService;
+
+    @Test
+    public void shouldBeRestoredInstantly() throws IOException, OperationException {
+        DocumentModel documentModel = createFileDocument(session, true);
+        // first make the move to cold storage
+        moveContentToColdStorage(session, documentModel);
+        session.saveDocument(documentModel);
+        restoreContentFromColdStorage(documentModel);
+    }
+
+    @Test
+    public void shouldFailRestoreNoColdStorageContent() throws OperationException, IOException {
+        DocumentModel documentModel = createFileDocument(session, true);
+        try {
+            // request a restore from the cold storage
+            restoreContentFromColdStorage(documentModel);
+            fail("Should fail because there no cold storage content associated to this document.");
+        } catch (NuxeoException e) {
+            assertEquals(SC_CONFLICT, e.getStatusCode());
+        }
+    }
+
+    protected void restoreContentFromColdStorage(DocumentModel documentModel) throws OperationException, IOException {
+        try (OperationContext context = new OperationContext(session)) {
+            context.setInput(documentModel);
+            automationService.run(context, RestoreFromColdStorage.ID);
+
+            waitForAsyncCompletion();
+            documentModel.refresh();
+            // check document
+            assertFalse(documentModel.hasFacet(ColdStorageHelper.COLD_STORAGE_FACET_NAME));
+
+            // check blobs
+            Blob fileContent = (Blob) documentModel.getPropertyValue(ColdStorageHelper.FILE_CONTENT_PROPERTY);
+            assertEquals(FILE_CONTENT, fileContent.getString());
+
+            // check the restore notification
+            String username = NotificationConstants.USER_PREFIX + session.getPrincipal().getName();
+            List<String> subscriptions = notificationManager.getSubscriptionsForUserOnDocument(username,
+                    documentModel);
+            assertTrue(subscriptions.contains(ColdStorageHelper.COLD_STORAGE_CONTENT_RESTORED_NOTIFICATION_NAME));
+            assertFalse(subscriptions.contains(ColdStorageHelper.COLD_STORAGE_CONTENT_AVAILABLE_NOTIFICATION_NAME));
+        }
+    }
+
+    protected void nextTransaction() {
+        TransactionHelper.commitOrRollbackTransaction();
+        TransactionHelper.startTransaction();
+    }
```

Comment:
I think we can remove this method and use the `transactionalFeature.nextTransaction();` we already have some example in the code.

Reply:
Also as we are using async event even for restore, it's worth to add the check on the fired events have a look at https://github.com/nuxeo/nuxeo-coldstorage/blob/master/nuxeo-coldstorage/src/test/java/org/nuxeo/coldstorage/helpers/AbstractTestColdStorageHelper.java#L271